src/hotspot/share/gc/shenandoah/shenandoahCardStats.hpp

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2022, Amazon.com, Inc. or its affiliates.  All rights reserved.
+ *
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_GC_SHENANDOAH_SHENANDOAHCARDSTATS_HPP
+#define SHARE_GC_SHENANDOAH_SHENANDOAHCARDSTATS_HPP
+
+#include "gc/shared/gc_globals.hpp"
+#include "gc/shenandoah/shenandoahNumberSeq.hpp"
+
+class ShenandoahCardStats: public CHeapObj<mtGC> {
+private:
+  size_t _cards_in_cluster;
+  HdrSeq* _local_card_stats;
+
+  bool _last_dirty;
+  bool _last_clean;
+
+  size_t _dirty_card_cnt;
+  size_t _clean_card_cnt;
+
+  size_t _dirty_run;
+  size_t _clean_run;
+
+  size_t _max_dirty_run;
+  size_t _max_clean_run;
+
+  size_t _dirty_obj_cnt;
+  size_t _clean_obj_cnt;
+
+  size_t _dirty_scan_cnt;
+  size_t _clean_scan_cnt;
+
+  size_t _alternation_cnt;
+
+public:
+  ShenandoahCardStats(size_t cards_in_cluster, HdrSeq* card_stats) :
+    _cards_in_cluster(cards_in_cluster),
+    _local_card_stats(card_stats),
+    _last_dirty(false),
```

Comment:
Should it always be the case that `_last_dirty != _last_clean`? Could we use one variable here instead of two?

Reply:
I believe we switch into one of two modes based on the first card we encounter. So there are 3 states: an initial (neither), and then subsequently either dirty or clean. So there are 3 states, which is 2 bits. It's possible I could shrink it to 2 bits by some cleverness, but figured I wouldn't try too hard as this is still all non-product. I'll think some more about it.

Reply:
I re-examined the code with an eye to reducing the use of flags. Although I think it might be possible, I believe it would complicate the structure of the code a bit because of teh asymmetric treatment of clean and dirty. As things stand the symmetry of the treatment leads to more uniform and consistent code structure that makes the code more maintainable. We can revisit the reduction in state separately in the fullness of time.

I'll resolve this comment based on the above reasoning.

Reply:
Okay, I was thinking that code of the form:
```
if (_last_dirty) {
  // ...
} else if (_last_clean) {
 // ...
}
```
Could be:
```
if (_last_dirty) {
  // ...
} else {
 // ...
}
```
But, as you point out, this doesn't handle the initial case.